### PR TITLE
COMMERCE-5449 | COMMERCE-5450 Inline editing and entity creation via dataset display

### PR DESCRIPTION
**Pull request by FabioDiegoMastrorilli**
Data Set Display should allow users to create new entities or editing existing ones directly from the component itself.

Requirements for editing:

1. The Editing mode can be activated via the action dropdown menu
2. The Editing mode can also be configured to be always on every item at the same time
3. Any column can be configured to switch the data renderer when the edit mode is active
4. The update request can have a default body object in case the API that handles the editing requires a richer payload. This default object properties will be appended to every update request.
5. The new data renderers input should handle: plain text, numbers, dates, booleans (via checkbox)

Requirements for adding: 

1. When the "Add new entity mode" is active, the table view will always start with a row with empty input boxes in it
2. The empty boxes should match the ones in the editing mode
3. As per the Editing mode. the post request can have a default body object in case the API that handles the entity creation requires a richer payload.
4. As per the Editing mode, the dataset should handle: plain text, numbers, dates, booleans (via checkbox)

Example for item creation and editing (alwayOn true): 
image

Example for item editing (alwayOn not true):
image
image

### CODE REVIEW

**Comment by jbalsas** [comment]
Thanks a lot, @FabioDiegoMastrorilli, looks nice, but feels like it might take a little bit to review! 😂 

Going to assign this to @bryceosterhaus for monday, but @markocikos, @julien, @carloslancha, feel free to take a look too, please! 🙏

**Comment by julien** [comment]
Going to have a look at the code changes.

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/DatasetDisplay.js [hunk 1]
```diff
@@ -221,10 +228,10 @@ const fluidDataSetDisplayProps = {
 					</>
 				);
 			},
-			icon: 'merge',
-			id: 'custom-table',
 			label: "Hey you don't know me",
```
Comment (julien) [inline modules/apps/commerce/commerce-frontend-js/test/dev/components/DatasetDisplay.js:231]: This looks like "random text" to test if something is working so I'm not sure, I'd suggest changing the value (I'm saying this with no context just looking at the code changes)

File: modules/apps/commerce/commerce-frontend-js/test/dev/components/DatasetDisplay.js [hunk 2]
```diff
@@ -546,7 +553,7 @@ const productsDataSetDisplayProps = {
 		{
 			id: 'blbl',
 			label: 'Custom Filter',
-			moduleUrl: '/blblasd/asd/basdkj',
+			moduleURL: '/blblasd/asd/basdkj',
```
Comment (julien) [inline modules/apps/commerce/commerce-frontend-js/test/dev/components/DatasetDisplay.js:556]: This also looks like a "test" value, so I would also either set it as an empty string or an actual URL.

File: modules/apps/commerce/commerce-frontend-js/test/dev/public/js/static-env-utils.js [hunk 3]
```diff
@@ -45,11 +45,22 @@ window.Liferay = {
 	on: (name, fn) => {
 		window.addEventListener(name, fn);
 	},
-	staticEnvHeaders: new Headers({
+};
+
+window.defaultFetch = fetch;
+
+window.fetch = (resource, {headers, ...init}) => {
```
Comment (julien) [inline modules/apps/commerce/commerce-frontend-js/test/dev/public/js/static-env-utils.js:52]: Couldn't you use [`fetch` from `frontend-js-web`](https://github.com/liferay/liferay-portal/blob/cb2b504c40d7a086dd7fa18f183b016ce2dd2347/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/fetch.es.js) and pass those `Headers`?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js [hunk 4]
```diff
@@ -75,29 +78,30 @@ function DataSetDisplay({
...
-	const [dataLoading, setDataLoading] = useState(false);
 	const [componentLoading, setComponentLoading] = useState(false);
-	const [dataSetDisplaySupportSidePanelId] = useState(
-		sidePanelId || 'support-side-panel-' + getRandomId()
-	);
-
+	const [dataLoading, setDataLoading] = useState(!!apiURL);
 	const [dataSetDisplaySupportModalId] = useState(
 		'support-modal-' + getRandomId()
 	);
```
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:88]: Couldn't we use (template literals) instead of concatenation?

```js
const [dataSetDisplaySupportModalId] = useState(
  `support-modal-${getRandomId()}`
);
```

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/data_renderers/InputDateTimeRenderer.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		dateInstance.getFullYear() +
+		'-' +
+		('0' + (dateInstance.getMonth() + 1)).slice(-2) +
+		'-' +
+		('0' + dateInstance.getDate()).slice(-2) +
+		'T' +
+		('0' + dateInstance.getHours()).slice(-2) +
+		':' +
+		('0' + dateInstance.getMinutes()).slice(-2);
+
```
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/data_renderers/InputDateTimeRenderer.js:32]: I'd also use template literals here instead of concatenation

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/utilities/dataRenderers.js [hunk 6]
```diff
@@ -12,9 +12,19 @@
  * details.
  */
 
-import dataRenderers from '../data_renderers/index';
+import dataRenderers, {inputRenderers} from '../data_renderers/index';
```
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/utilities/dataRenderers.js:15]: Just a question, I see we have a `utilities` directory (or package), is there any reason why it's called like that and not `utils`?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/entry.js [hunk 7]
```diff
@@ -13,25 +13,25 @@
...
 import {AppContext} from './AppContext';
 import DataSetDisplay from './DataSetDisplay';
 import ViewsContext, {viewsReducer} from './views/ViewsContext';
 
 const App = ({apiURL, appURL, portletId, ...props}) => {
-	const {
-		activeViewSettings: {name: activeViewName, visibleFieldNames = {}},
-		views,
-	} = props;
+	const activeViewName = props.activeViewSettings?.name;
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/entry.js:24]: We can get `activeViewSettings` and `views` destructured from props instead use `props.view` etc

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import {
+	getDataRendererById,
+	getDataRendererByURL,
+} from '../../utilities/dataRenderers';
+
+function InlineEditInputRenderer({type, value, ...otherProps}) {
+	const {modifiedItems, updateItem} = useContext(DataSetDisplayContext);
+	let InputRenderer;
+
+	switch (type) {
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js]: This could be done inside a `useEffect` so it's only calculated if `type` changes

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+			InputRenderer = InputCheckboxRenderer;
+			break;
+		default:
+			InputRenderer = InputTextRenderer;
+	}
+
+	let inputValue = value;
+
+	const formattedFieldName = Array.isArray(otherProps.options.fieldName)
+		? otherProps.options.fieldName[0]
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js:34]: add `options`  as a destructured prop instead getting it from `otherProps`

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		: otherProps.options.fieldName;
+
+	if (
+		modifiedItems[otherProps.itemId] &&
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableCell.js]: same as `options` and `otherProps`

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	fieldName,
+	hideColumnLabel,
+	label,
+	sortable,
+	sorting,
+	sortingKey: sortingKeyProp,
+	updateSorting,
+}) {
+	const sortingKey =
+		sortingKeyProp || (Array.isArray(fieldName) ? fieldName[0] : fieldName);
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:34]: Feels weird that fieldName could be an array

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+	const sortingMatch = sorting.find((element) => element.key === sortingKey);
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:36]: Place this inside `useEffect` to only run it when `sorting` changes

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	return (
+		<ClayTable.Cell
+			className={classNames(
+				contentRenderer && `content-renderer-${contentRenderer}`,
+				expandableColumns
+					? expand && 'table-cell-expand-small'
+					: 'table-cell-expand-smaller'
+			)}
```
Comment (bryceosterhaus) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:71]: You can also pass an object to `classNames` and toggle the key(class to be applied), with a boolean in its value.

```js
classNames({
  [`content-renderer-${contentRenderer}`]: contentRenderer,
  'table-cell-expand-smaller': !expandableColumns,
  'table-cell-expand-small': expandableColumns && expand
})
```

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+			headingTitle
+		>
+			{sortable ? (
+				<a
+					className="inline-item text-nowrap text-truncate-inline"
+					data-senna-off
+					href="#"
+					onClick={handleSortingCellClick}
+				>
+					{(!hideColumnLabel && label) || ''}
```
Comment (bryceosterhaus) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:82]: I don't think the `|| ''` is necessary here... I may be mistaken though

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+					<span className="inline-item inline-item-after sorting-icons-wrapper">
+						<ClayIcon
+							className={classNames(
+								'sorting-icon',
+								sortingMatch?.direction === 'asc' && 'active'
```
Comment (bryceosterhaus) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:87]: Same as above, it's probably gonna be a bit more readable to use the object syntax for `classNames`. Note that you can use both an object and a string. So the first argument would be `'sorting-icon'` and the second arg would be an object for toggling `'active'`

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+							)}
+							draggable
+							symbol="order-arrow-up"
+						/>
+						<ClayIcon
+							className={classNames(
+								'sorting-icon',
+								sortingMatch?.direction === 'desc' && 'active'
```
Comment (bryceosterhaus) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:95]: Same as above comment

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+							)}
+							draggable
+							symbol="order-arrow-down"
+						/>
+					</span>
+				</a>
+			) : (
+				(!hideColumnLabel && label) || ''
```
Comment (bryceosterhaus) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:103]: Same as above, I'm not sure the `|| ''` is necessary

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js [hunk 18]
```diff
@@ -75,29 +78,30 @@ function DataSetDisplay({
 	sorting: sortingProp,
 	style,
 }) {
+	const {apiURL} = useContext(AppContext);
+
```
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:82]: I'd be *very careful* about adding `useCallback` or `useMemo` just to avoid "re-creating functions repeatedly". Note that in order to make use of either, you are instantiating a new function every time through your render anyway; eg:

```js
```

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:82]: Whoops, accidentally focused the "Comment" button and hit "Enter" before I finished entering the comment. Updated comment incoming...

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:82]: I'd be *very careful* about adding `useCallback` or `useMemo` just to avoid "re-creating functions repeatedly". Note that in order to make use of either, you are instantiating a new function every time through your render anyway; eg: (function chosen at random, but it's just for illustration):

```js
function DataSetDisplay() {
    // stuff

    function updateDataSetItems(dataSetData) {
        setTotal(dataSetData.totalCount);
        updateItems(dataSetData.items);
    }

    // etc
}
```

vs:

```js
function DataSetDisplay() {
    // stuff

    const updateDataSetItems = useCallback((dataSetData) => {
        setTotal(dataSetData.totalCount);
        updateItems(dataSetData.items);
    }, [...]);

    // etc
}
```

The first version is re-instantiating a function object on every render, but so is the second one (the arrow function). In addition, it is paying the overhead of _calling_ `useCallback` on every render, and it is allocating a dependencies array on every render. `useCallback` itself is going to incur the cost of comparing the items in the dependency array (using `Object.is()`), as well as the memory overhead of any related book-keeping details that React is maintaining internally.

Now, I'm not saying "never use `useCallback` or `useMemo`"; I'm just saying, make sure you have a good reason to use them. A non-exhaustive list of good reasons for using them is:

- Use `useMemo` to avoid repeatedly recomputing an intermediate value that is expensive to evaluate (note: in order for this to pay off, the value shouldn't be changing on every render _and_ the cost of determining equality should be cheaper than just redoing the calculation — remember that you're paying the cost of the equality comparison plus the overhead of the function call into React internals plus whatever internal book-keeping React might be doing). In many cases, it's cheaper to just re-do the computation every time.
- Use `useCallback` when you want to maintain a stable identity for a callback function that you will be passing down to children; this can be important when the callback ends up being a member of a `useEffect` dependencies array in the child (or grandchild, or great-grandchild etc). If the callback hasn't really changed, you may find that failure to use `useCallback` causes children to re-run effects unnecessarily because they think something has changed when it really hasn't. This is another context-dependent thing: sometimes you know everything about your children subtree and you can know deterministically that `useCallback` _is_ required; at other times you are writing "library" code and you have no idea what your children are going to be doing, so you may want to err on the side of being defensive and using `useCallback` "just in case".

Another angle on this: https://kentcdodds.com/blog/usememo-and-usecallback/

> _Performance optimizations ALWAYS come with a cost but do NOT always come with a benefit._

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:82]: Made this issue for us to put something about these hooks in the guidelines: https://github.com/liferay/liferay-frontend-projects/issues/383

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/utilities/dataRenderers.js (hunk 6 above)
Comment (FabioDiegoMastrorilli) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/utilities/dataRenderers.js:15]: Actually not so I'll proceed with renaming it! :)

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js [hunk 19]
```diff
@@ -423,10 +423,82 @@ function DataSetDisplay({
 		});
 	}
 
+	function updateItem(itemKey, property, value = null) {
+		const formattedProperty = Array.isArray(property)
```
[the remark this answers is not in the document]
Comment (FabioDiegoMastrorilli) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:431]: When we define a property to be displayed in a column, we might need to get some nested value: in that case we pass an array with the keys, like: `['statusWorkFlowInfo', 'label_i18n']` or `['name', 'en_US']`

So far, the data set uses only the first array item as a body key when the POST get submitted. We can actually add a flag in the column schema to decide wether we wanna reconstruct the nested items structure or not

[the remark this answers is not in the document]
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js:431]: In that case this could be more clear if `property` was an object with `name` an `nestedValues` or something like that (IMHO)

File: modules/apps/commerce/commerce-frontend-js/test/dev/public/js/static-env-utils.js (hunk 3 above)
Comment (FabioDiegoMastrorilli) [reply modules/apps/commerce/commerce-frontend-js/test/dev/public/js/static-env-utils.js:52]: Julien on this one I preferred keeping it as it is. This file contains some utilities we place a static env to replace standard liferay ones (like `Liferay.Language.get` or `themeDisplay.getLanguageId`) and it doesn't get compiled. 

Its only purpose is avoiding to write exceptions (that are aimed to make everything working in static env) in code which will be deployed

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js [hunk 20]
```diff
@@ -76,28 +77,27 @@ function DataSetDisplay({
...
 		'support-modal-' + getRandomId()
 	);
-
-	const [selectedItemsValue, setSelectedItemsValue] = useState(
-		selectedItems || []
+	const [dataSetDisplaySupportSidePanelId] = useState(
+		sidePanelId || 'support-side-panel-' + getRandomId()
+	);
+	const [delta, setDelta] = useState(
+		pagination.initialDelta || pagination.deltas[0].label
```
[the remark this answers is not in the document]
Comment (FabioDiegoMastrorilli) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js]: I set it in DataSetDisplay.defaultProps so it's pretty safe I hope

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js (hunk 15 above)
Comment (FabioDiegoMastrorilli) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:95]: Only on this one and the other one below I would prefer keeping them as they are. Passing both a string and an object to `classnames` looks a bit less readable (or maybe I'm just not used to it). `active` also will not be highlighted as a string due to our linter an it looks a bit strange to me:

```
className={classNames(
	'sorting-icon',
	{
		active: sortingMatch?.direction === 'asc'
	}
)}
```

Btw if there are some standards we have to follow in this cases, not problems at all :)

**Comment by FabioDiegoMastrorilli** [comment]
@wincent @bryceosterhaus @carloslancha @julien the PR has been updated! Thanks a lot for your code review! :D 

cc @jbalsas

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js (hunk 15 above)
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:95]: I don't think we have any standards written down, and we probably won't either unless we start to see `classNames()` usage "spiralling into chaos" 😂.

I don't have an opinion on this strong enough to make me want to comment, but seeing as the thread is already open, the way I usually tend to write these things to avoid the feeling of impedance mismatch you get when mixing string and object is to just use object for everything:

```jsx
<ClayIcon
    className={{
        active: sortingMatch?.direction === 'asc',
        'sorting-icon': true,
    }}
/>
```

but like all things, it's a trade-off; like you say @FabioDiegoMastrorilli, you lose the string highlighting of the `active` classname. We should be able to get it back though because I believe a long-standing Prettier bug has been fixed (see [discussion](https://github.com/liferay/eslint-config-liferay/pull/39)) that would format the above as:

```js
<ClayIcon
    className={{
        'active': sortingMatch?.direction === 'asc',
        'sorting-icon': true,
    }}
/>
```

Comment (bryceosterhaus) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/views/table/TableHeadCell.js:95]: @FabioDiegoMastrorilli, thats fine, its not vital that you use the object notation, I just wanted to make sure you were aware of it.

**Comment by wincent** [comment]
@FabioDiegoMastrorilli: I rebased this to resolve the conflicts (in the lang files, so I just re-ran `buildLang`) and pushed to your fork. I gather this has been reviewed enough now by multiple pairs of eyes so going to see if it forwards.

**Comment by wincent** [comment]
ci:forward

**Comment by wincent** [comment]
ci:stop

**Comment by wincent** [comment]
(Stopped because `buildLang` got all confused by the merge conflict markers and inserted a bunch of Chinese(?) into every file.)

**Comment by wincent** [comment]
ci:forward
